### PR TITLE
feat(button): add ionic theme implementation

### PR DESCRIPTION
**Pull request by joselrio**
Issue number: **ROU-4815**

---------

## What is the new behavior?

- Adds an ionic theme style implementation for `ion-button`.
  - Adds a new `rectangular` shape to the button.
  - Adds two new button sizes: `xsmall` and `xlarge`.
  - Adds a new `focus-ring` CSS shadow part for the ionic theme.

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Preview

Screen-Recording-2024-03-20-at-1

### CODE REVIEW

File: core/src/components/button/button.tsx
Comment (OS-giulianasilva) [inline core/src/components/button/button.tsx:126]:
```suggestion
   * Option `"xsmall"` and `"xlarge"` are only available for Ionic theme.
```

File: core/src/components/button/button.tsx
Comment (OS-giulianasilva) [inline core/src/components/button/button.tsx:127]:
```suggestion
   * At ios and md, if they are in use undefined will be assumed.
```

**Review by OS-giulianasilva — CHANGES_REQUESTED** [review]

**Review by bmarcelino-fe — APPROVED** [review]

**Review by OS-giulianasilva — CHANGES_REQUESTED** [review]
The button sizes seem to not be exactly what is designed, for example, the Extra large button height is 70px instead of 56px.

File: core/src/themes/ionic.functions.sizes.scss
Comment (joselrio) [reply core/src/themes/ionic.functions.sizes.scss:11]: Due a11y we should always use it in order to let user set their own size if they have visual disabilities. Once font-size was already set in `em` I decided to set everything in `em` this way everything will be scaled based on browser setting value ;)

Comment (joselrio) [reply core/src/themes/ionic.functions.sizes.scss]: @OS-giulianasilva I made confusion between `em` and `rem` 🤦 ...
Also design doc was not properly ok.
That's fixed 😉

Again, thank you for raising it! :)

File: core/src/components/button/button.ionic.scss
Comment (sean-perkins) [inline core/src/components/button/button.ionic.scss:22]: Ionic Framework includes a set of Sass mixins in favor of directional properties that can change from LTR → RTL. 

Here, we should use the `@include position(50%, 0, 0, 0);` mixin. It will align with code standards in other style files and remove the need for disabling stylelint.

File: core/src/components/button/button.ionic.scss
Comment (sean-perkins) [inline core/src/components/button/button.ionic.scss:44]: This feedback applies to all usages of padding/margin used in the implementation. 

Let's use the `@include margin` and `@include padding` mixins.

File: core/src/components/button/button.tsx
Comment (sean-perkins) [inline core/src/components/button/button.tsx:126]: Suggestion:

> `"xsmall"` is only available for the `"ionic"` theme.
> `"xlarge"` is only available for the `"ionic"` theme. 

I don't believe we need to specify what happens on iOS or MD here. I suspect we will later fill in values for those themes 👍

File: core/src/components/button/button.tsx
Comment (sean-perkins) [inline core/src/components/button/button.tsx:23]:
```suggestion
 * @part focus-ring - The visual indicator that appears as an outline around the button when focused. Only available for the Ionic theme.
```

File: core/src/components/button/test/theme-ionic/index.html
Comment (sean-perkins) [inline core/src/components/button/test/theme-ionic/index.html:2]: You can add `theme="ionic"` to the `html` node to avoid having to specify it on any children elements.

File: core/src/themes/ionic.functions.sizes.scss
Comment (sean-perkins) [inline core/src/themes/ionic.functions.sizes.scss:12]: Is there any difference between the use of `calc` here vs. the `math.div` function from `sass:math`? 

We have a similar utility for dynamic font and I was curious if there was a reason to use one approach vs. the other. 

I do think this utility is helpful to have 👍

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (sean-perkins) [inline core/src/components/button/test/theme-ionic/button.e2e.ts:38]:
```suggestion
configs({ themes: ['ionic'], modes: ['md'] }).forEach(({ config, screenshot, title }) => {
```

Since the appearance and behavior we are testing for the button is consistent between platforms, we can limit the testing to a specific mode.

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (sean-perkins) [inline core/src/components/button/test/theme-ionic/button.e2e.ts:34]: Capturing an entire page screenshot will result in additional diffs in the future (from changes not specific to the button) and significantly larger file sizes. 

Can we either:
1. Capture just the individual elements (smallest file size with this approach)
2. Capture a screenshot of the "Preview" and "States" containers?

This will remove the "Preview Options" section and header from being captured in the screenshot.

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (sean-perkins) [inline core/src/components/button/test/theme-ionic/button.e2e.ts:1]: I would recommend taking a look at: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/button/test/size/button.e2e.ts to model the tests around. 

The tests will run significantly faster with a much smaller file size as a result.

File: core/src/components/button/button.tsx
Comment (sean-perkins) [inline core/src/components/button/button.tsx:44]: Is the circle button required for this initial work? We have in progress work to implement this: https://github.com/ionic-team/ionic-framework/pull/29161.

I believe later we could add the circle button implementation for the ionic theme, after the feature is approved/merged.

**Review by sean-perkins — COMMENTED** [review]
Tested locally, really great work here! 

I've commented on some code style and testing best practices. 

When making changes to some of the code documentation, you will need to run `npm run build` locally to regenerate certain files. 

I need to run, but I can provide information on generating screenshot in a follow-up comment later.

File: core/src/themes/ionic.functions.sizes.scss
Comment (OS-giulianasilva) [reply core/src/themes/ionic.functions.sizes.scss]: Thank you for the explanation :)

**Review by OS-giulianasilva — APPROVED** [review]

File: core/src/components/button/button.tsx
Comment (joselrio) [reply core/src/components/button/button.tsx]: Indeed! This line was removed...

File: core/src/components/button/test/theme-ionic/index.html
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/index.html:2]: I used the theme="ionic" at the ion-button directly since we do not have ionic theme styles for the header yet. This way screen will have a better appearance. That said, I would say once we get everything styled we can remove this inline theme attrs ;)

File: core/src/themes/ionic.functions.sizes.scss
Comment (joselrio) [reply core/src/themes/ionic.functions.sizes.scss]: There's not, it's the same since sass has the fallback to use calc at the same way browser do. didn't notice the usage of the `math.div`, in order to keep consistency I updated it to use `math.div` :)

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/button.e2e.ts]: Done!

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/button.e2e.ts]: Done!

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/button.e2e.ts:1]: Done!

File: core/src/components/button/button.tsx
Comment (joselrio) [reply core/src/components/button/button.tsx:44]: I did already that question as well, but it was at the design docs and also at the button requirements list, that's way I implemented it :/

**Comment by joselrio** [comment]
@sean-perkins amazing feedback! Super useful to get deeper context... 
Thank you! :)

File: core/src/themes/ionic.functions.sizes.scss
Comment (brandyscarney) [inline core/src/themes/ionic.functions.sizes.scss:11]:
```suggestion
@function px-to-rem($pixels, $context: $browser-context) {
```

Could we change this case to match the functions in other files:
- https://github.com/ionic-team/ionic-framework/blob/main/core/src/themes/ionic.functions.color.scss#L57
- https://github.com/ionic-team/ionic-framework/blob/main/core/src/themes/ionic.functions.string.scss#L7

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:39]:
```suggestion
    @include margin-horizontal(null, pxToRem(8));
```

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:43]:
```suggestion
    @include margin-horizontal(pxToRem(8), null);
```

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:55]:
```suggestion
    @include margin-horizontal(null, pxToRem(4));
```

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:59]:
```suggestion
    @include margin-horizontal(pxToRem(4), null);
```

**Review by brandyscarney — CHANGES_REQUESTED** [review]

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:165]:
```suggestion

```

We should actually remove this and set the `--background-activated` and `--color-activated` variables inside of `.button-solid`, `.button-outline` and `.button-clear`. This is my fault for including this in the design doc.

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:201]: This value is incorrect. According to Figma it should match the Ghost (Clear) button:

Screenshot 2024-03-21 at 12 07 33 PM

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:206]:
```suggestion
:host(.button-outline) {
  --border-width: #{$button-ionic-outline-border-width};
  --border-style: #{$button-ionic-outline-border-style};
  --background-activated: #e3e3e3;
  --background-activated-opacity: 1;
  --background-focused: transparent;
  --background-hover: transparent;
  --background-focused-opacity: .1;
  --color-activated: #{ion-color(primary, base)};
}

:host(.button-outline.ion-focused) {
  --border-width: 0;
}
```

I might be wrong but I think these are the values for outline to match the Figma.

File: core/src/components/button/button.ionic.scss
Comment (brandyscarney) [inline core/src/components/button/button.ionic.scss:215]:
```suggestion
:host(.button-clear) {
  --background-activated: #e3e3e3;
  --background-activated-opacity: 1;
  --background-focused: transparent;
  --background-hover: transparent;
}
```

File: core/src/themes/ionic.functions.sizes.scss
Comment (joselrio) [reply core/src/themes/ionic.functions.sizes.scss]: Done!

File: core/src/components/button/button.ionic.vars.scss
Comment (joselrio) [reply core/src/components/button/button.ionic.vars.scss]: Done!

File: core/src/components/button/button.ionic.scss
Comment (joselrio) [reply core/src/components/button/button.ionic.scss]: Done!

**Comment by joselrio** [comment]
@brandyscarney Thank you on being that meniscus on raising the misalignments I had! 💪

File: core/src/components/button/test/theme-ionic/index.html
Comment (thetaPC) [inline core/src/components/button/test/theme-ionic/index.html:1]: We should be using the existing test files instead of making a new one specific to a theme. We can use the query `?ionic:theme=ionic` to view the changes being made to button. By sticking to the existing file structure, we have less files to maintain.

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (thetaPC) [inline core/src/components/button/test/theme-ionic/button.e2e.ts:1]: I also would push to have these tests moved to their appropriate folders. We already have a test folder for sizes, outline, clear, etc. I would recommend making a new folder for shapes.

File: core/src/components/button/button.ionic.scss
Comment (thetaPC) [inline core/src/components/button/button.ionic.scss:230]:
```suggestion
}

```
Missing new line.

**Review by thetaPC — CHANGES_REQUESTED** [review]

File: core/src/components/button/test/theme-ionic/button.e2e.ts
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/button.e2e.ts:1]: @sean-perkins did a [PR](https://github.com/ionic-team/ionic-framework/pull/29201#pullrequestreview-1954566017) for it already :)

File: core/src/components/button/test/theme-ionic/index.html
Comment (joselrio) [reply core/src/components/button/test/theme-ionic/index.html:1]: @sean-perkins did a [PR](https://github.com/ionic-team/ionic-framework/pull/29201#pullrequestreview-1954566017) for it already :)

File: core/src/components/button/button.tsx
Comment (joselrio) [reply core/src/components/button/button.tsx]: I forgot to remove it! 🤦 
Removed!

File: core/src/components/button/button.tsx
Comment (joselrio) [reply core/src/components/button/button.tsx]: I would say to keep it for now since we've at the requirements and design doc as well. Once we've that clear we can remove it and use the new one if that's the expected ;)

Comment (thetaPC) [reply core/src/components/button/button.tsx]: The `next` branch has been updated to include the circle button. So we can resolve this conversation.

**Review by thetaPC — CHANGES_REQUESTED** [review]
Request me once Sean's [PR](https://github.com/ionic-team/ionic-framework/pull/29201) has been merged into this branch.

**Comment by joselrio** [comment]
@thetaPC missing tests will be done under another task that will be created. This PR should not be blocked for now due to those missing tests. Thank you :)

**Review by sean-perkins — APPROVED** [review]
Great work on this 👏 

I've updated the use of the `border-radius` property for our `@include border-radius` mixin. I've also updated a few comment blocks in the Sass stylesheet, but we can continue to update that as additional features are added.

I believe this PR is ready for merge. I've linked technical debt tasks after another PR for test infrastructure is merged. When that happens, a test template will be removed and new screenshots for test cases will be added.

We should provide the Design team with a build of these changes, so they can validate the accuracy of the implementation as well, unless they have already looked and approved the implementation separately.

**Comment by thetaPC** [comment]
@joselrio I noticed that Sean's PR was merged into this one. But I'm not seeing his changes here.

**Comment by sean-perkins** [comment]
@thetaPC the changes were added as part of this commit: https://github.com/ionic-team/ionic-framework/pull/29187/commits/7bfb4833729d18c2c176467ddf6a0475c3b5324d. I generated the screenshots and added a few test cases that are not blocked by #29206.

I created FW-6077 to track adding the remaining tests. I added `// TODO` statements in the code where those apply.

**Pull request merged**
